File: app/src/main/java/com/github/dedis/student20_pop/model/network/method/Unsubscribe.java
package com.github.dedis.student20_pop.model.network.method;

import com.github.dedis.student20_pop.utility.protocol.MessageHandler;

/**
 * Unsubscribe from a channel
 */
public final class Unsubscribe extends Query {

    public Unsubscribe(String channel, int id) {
        super(channel, id);
    }

    @Override
    public void accept(MessageHandler handler) {
        handler.handle(this);
    }

    @Override
    public String getMethod() {
        return Method.UNSUBSCRIBE.getMethod();
    }
}

File: be1-go/internal/handler/method/unsubscribe/unsubscribe.go
package unsubscribe

import (
	"encoding/json"
	"popstellar/internal/errors"
	"popstellar/internal/handler/messagedata/root"
	"popstellar/internal/message/query/method"
	"popstellar/internal/network/socket"
	"popstellar/internal/repository"
)

type Handler struct {
	subs repository.SubscriptionManager
}

func New(subs repository.SubscriptionManager) *Handler {
	return &Handler{subs: subs}
}

func (h *Handler) Handle(socket socket.Socket, msg []byte) (*int, error) {
	var unsubscribe method.Unsubscribe
	err := json.Unmarshal(msg, &unsubscribe)
	if err != nil {
		return nil, errors.NewJsonUnmarshalError(err.Error())
	}

	if root.Root == unsubscribe.Params.Channel {
		return &unsubscribe.ID, errors.NewAccessDeniedError("cannot Unsubscribe from root channel")
	}

	err = h.subs.Unsubscribe(unsubscribe.Params.Channel, socket)
	if err != nil {
		return &unsubscribe.ID, err
	}

	socket.SendResult(unsubscribe.ID, nil, nil)

	return &unsubscribe.ID, nil
}
